cpp/src/parquet/types.h

```diff
@@ -89,11 +90,831 @@ struct LogicalType {
+    JSON,
+    BSON,
+    UUID,
+    None
+  };
+
+  enum class TimeUnit { Unknown = 0, MilliSeconds = 1, MicroSeconds, NanoSeconds };
+
+#define time_unit_string(u___)                                                      \
+  ((u___) == LogicalAnnotation::TimeUnit::MilliSeconds                              \
+       ? "milliseconds"                                                             \
+       : ((u___) == LogicalAnnotation::TimeUnit::MicroSeconds                       \
+              ? "microseconds"                                                      \
+              : ((u___) == LogicalAnnotation::TimeUnit::NanoSeconds ? "nanoseconds" \
+                                                                    : "unknown")))
+
+  // If possible, return an annotation equivalent to the given legacy converted type
+  // (and decimal metatdata if applicable).
+  static std::shared_ptr<const LogicalAnnotation> FromConvertedType(
+      const LogicalType::type converted_type,
+      const schema::DecimalMetadata converted_decimal_metadata = {false, -1, -1});
+
+  // Return the annotation represented by the Thrift intermediary object
+  static std::shared_ptr<const LogicalAnnotation> FromThrift(
+      const format::LogicalType& thrift_logical_type);
+
+  // Return the requested type of annotation ...
+  static std::shared_ptr<const LogicalAnnotation> UnknownType(void);
+  static std::shared_ptr<const LogicalAnnotation> StringType(void);
+  static std::shared_ptr<const LogicalAnnotation> MapType(void);
+  static std::shared_ptr<const LogicalAnnotation> ListType(void);
+  static std::shared_ptr<const LogicalAnnotation> EnumType(void);
+  static std::shared_ptr<const LogicalAnnotation> DecimalType(int32_t precision);
+  static std::shared_ptr<const LogicalAnnotation> DecimalType(int32_t precision,
+                                                              int32_t scale);
+  static std::shared_ptr<const LogicalAnnotation> DateType(void);
+  static std::shared_ptr<const LogicalAnnotation> TimeType(
+      bool adjusted_to_utc, LogicalAnnotation::TimeUnit time_unit);
+  static std::shared_ptr<const LogicalAnnotation> TimestampType(
+      bool adjusted_to_utc, LogicalAnnotation::TimeUnit time_unit);
+  static std::shared_ptr<const LogicalAnnotation> IntervalType(void);
+  static std::shared_ptr<const LogicalAnnotation> IntType(int bit_width, bool is_signed);
+  static std::shared_ptr<const LogicalAnnotation> NullType(void);
+  static std::shared_ptr<const LogicalAnnotation> JSONType(void);
+  static std::shared_ptr<const LogicalAnnotation> BSONType(void);
+  static std::shared_ptr<const LogicalAnnotation> UUIDType(void);
+  static std::shared_ptr<const LogicalAnnotation> NoType(void);
+
+  // Is this annotation consistent with the given underlying physical type?
+  inline virtual bool is_applicable(const parquet::Type::type primitive_type,
+                                    const int32_t primitive_length = -1) const {
+    return true;
```

Comment:
This method is now pure virtual on the base class.  The new UnknownAnnotation class (which now serves as the NA type instead of the base class) answers vacuously true here and I inserted a comment to that effect.